File: IskrClockiOS/IskrClockiOS/Services/TimerManager.swift
//
//  TimerManager.swift
//  IskrClockiOS
//

import Foundation
import Combine

class TimerManager: ObservableObject {
    @Published var timerState = TimerState()
    private var timer: Timer?

    func start() {
        timerState.start()
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.timerState.tick()
            if self?.timerState.remainingSeconds == 0 {
                self?.timerCompleted()
            }
        }
    }

    func pause() {
        timerState.pause()
        timer?.invalidate()
    }

    func resume() {
        timerState.resume()
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.timerState.tick()
            if self?.timerState.remainingSeconds == 0 {
                self?.timerCompleted()
            }
        }
    }

    func reset() {
        timer?.invalidate()
        timerState.reset()
    }

    private func timerCompleted() {
        timer?.invalidate()
        timerState.isRunning = false

        // Play notification sound
        AudioPlayer.shared.playClassicAlarm()

        // Send notification
        NotificationManager.shared.sendAlarmNotification()
    }
}

class StopwatchManager: ObservableObject {
    @Published var stopwatchState = StopwatchState()
    private var timer: Timer?

    func start() {
        stopwatchState.start()
        timer = Timer.scheduledTimer(withTimeInterval: 0.01, repeats: true) { [weak self] _ in
            self?.stopwatchState.update()
        }
    }

    func pause() {
        stopwatchState.pause()
        timer?.invalidate()
    }

    func resume() {
        stopwatchState.resume()
        timer = Timer.scheduledTimer(withTimeInterval: 0.01, repeats: true) { [weak self] _ in
            self?.stopwatchState.update()
        }
    }

    func reset() {
        timer?.invalidate()
        stopwatchState.reset()
    }

    func lap() {
        stopwatchState.lap()
    }
}

File: IskrClockFlutter/lib/services/timer_manager.dart
import 'dart:async';
import 'package:flutter/material.dart';
import '../models/timer_state.dart';
import 'audio_player_service.dart';
import 'notification_service.dart';

class TimerManager with ChangeNotifier {
  TimerState _timerState = TimerState();
  Timer? _timer;
  final AudioPlayerService _audioPlayer = AudioPlayerService();
  final NotificationService _notificationService = NotificationService();

  TimerState get timerState => _timerState;

  void start() {
    _timerState.start();
    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      _timerState.tick();
      if (_timerState.remainingSeconds == 0) {
        _timerCompleted();
      }
      notifyListeners();
    });
    notifyListeners();
  }

  void pause() {
    _timerState.pause();
    _timer?.cancel();
    notifyListeners();
  }

  void resume() {
    _timerState.resume();
    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      _timerState.tick();
      if (_timerState.remainingSeconds == 0) {
        _timerCompleted();
      }
      notifyListeners();
    });
    notifyListeners();
  }

  void reset() {
    _timer?.cancel();
    _timerState.reset();
    notifyListeners();
  }

  void setTime({required int hours, required int minutes, required int seconds}) {
    _timerState.setTime(hours: hours, minutes: minutes, seconds: seconds);
    notifyListeners();
  }

  void _timerCompleted() {
    _timer?.cancel();
    _timerState.isRunning = false;

    // Play notification sound
    _audioPlayer._playClassicAlarm();

    // Send notification
    _notificationService.sendAlarmNotification();

    notifyListeners();
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }
}

class StopwatchManager with ChangeNotifier {
  StopwatchState _stopwatchState = StopwatchState();
  Timer? _timer;

  StopwatchState get stopwatchState => _stopwatchState;

  void start() {
    _stopwatchState.start();
    _timer = Timer.periodic(const Duration(milliseconds: 10), (timer) {
      _stopwatchState.update();
      notifyListeners();
    });
    notifyListeners();
  }

  void pause() {
    _stopwatchState.pause();
    _timer?.cancel();
    notifyListeners();
  }

  void resume() {
    _stopwatchState.resume();
    _timer = Timer.periodic(const Duration(milliseconds: 10), (timer) {
      _stopwatchState.update();
      notifyListeners();
    });
    notifyListeners();
  }

  void reset() {
    _timer?.cancel();
    _stopwatchState.reset();
    notifyListeners();
  }

  void lap() {
    _stopwatchState.lap();
    notifyListeners();
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }
}
